src/StageStridedLoads.cpp

```diff
@@ -0,0 +1,366 @@
+                return false;
+            } else if (scope < other.scope) {
+                return true;
+            } else if (scope > other.scope) {
+                return false;
+            } else if (allocation < other.allocation) {
+                return true;
+            } else if (allocation > other.allocation) {
+                return false;
+            } else if (type < other.type) {
+                return true;
+            } else if (other.type < type) {
+                return false;
+            } else if (buf < other.buf) {
+                return true;
+            } else if (buf > other.buf) {
+                return false;
+            } else {
+                return graph_less_than(base, other.base);
+            }
+        }
+    };
+    // Entry entry maps from an offset from the base to a vector of identical
+    // Load nodes with that offset.
+    std::map<Key, std::map<int64_t, std::vector<const Load *>>> found_loads;
+
+    // The current scope over which accesses definitely occur.
+    const IRNode *scope = nullptr;
+
+    Scope<const Allocate *> allocation_scope;
+
+    std::map<const IRNode *, const IRNode *> parent_scope;
+
+protected:
+    void visit(const Load *op) override {
+        if (is_const_one(op->predicate)) {
+            Expr idx = substitute_in_all_lets(simplify(common_subexpression_elimination(op->index)));
+            if (const Ramp *r = idx.as<Ramp>()) {
+                const int64_t *stride_ptr = as_const_int(r->stride);
+                int64_t stride = stride_ptr ? *stride_ptr : 0;
+                Expr base = r->base;
+                int64_t offset = 0;
+                const Add *base_add = base.as<Add>();
+                const int64_t *offset_ptr = base_add ? as_const_int(base_add->b) : nullptr;
+                if (offset_ptr) {
+                    base = base_add->a;
+                    offset = *offset_ptr;
+                }
+                // TODO: We do not yet handle nested vectorization here for
+                // ramps which have not already collapsed. We could potentially
+                // handle more interesting types of shuffle than simple flat slices.
+                if (stride >= 2 && stride < r->lanes && r->stride.type().is_scalar()) {
+                    const IRNode *s = scope;
+                    const Allocate *a = nullptr;
+                    if (allocation_scope.contains(op->name)) {
+                        a = allocation_scope.get(op->name);
+                    }
+                    found_loads[Key{op->name, base, stride, r->lanes, op->type, a, s}][offset].push_back(op);
+                }
+            }
+        }
+        IRVisitor::visit(op);
+    }
+
+    void visit(const For *op) override {
+        if (can_prove(op->extent > 0)) {
+            // The loop body definitely runs
+
+            // TODO: worry about different iterations of the loop somehow not
+            // providing the evidence we thought it did.
```

Comment:
(same as above) should this TODO be an issue? Or a blocker for this PR? I'm also not quite sure what the TODO means.